src/raglite/_extract.py

```diff
@@ -88,3 +97,275 @@ class MyNameResponse(BaseModel):
+            {
+                "key": "document_type",
+                "type": Literal["research-paper", "tutorial", "documentation"],
+                "prompt": "What type of document is this?",
+                "source": "content"  # Extract from document content
+            },
+            {
+                "key": "author_affiliation",
+                "type": Literal["academic", "industry", "government"],
+                "prompt": "What type of affiliation does this author likely have?",
+                "source": "author"  # Extract from existing 'author' metadata field
+            },
+            {
+                "key": "topics",
+                "type": list[Literal["AI", "ML", "NLP", "computer-vision", "robotics"]],
+                "prompt": "What topics does this document cover?",
+                "source": "content"  # Multi-choice field
+            },
+            {
+                "key": "summary",
+                "type": str,
+                "prompt": "Provide a brief summary of this document",
+                "source": "content",  # Extract from document content (default)
+                "max_chars": 2000  # Only analyze first 2000 characters
+            }
+        ]
+
+        # Expand metadata before inserting to database
+        expanded_docs = list(expand_document_metadata(documents, metadata_fields, raglite_config))
+
+        # Now insert to database
+        insert_documents(expanded_docs)
+    """
+    if not documents:
+        return
+
+    # Create dynamic Pydantic model for metadata validation.
+    fields: dict[str, Any] = {}
+    for field_spec in metadata_fields:
+        key = field_spec["key"]
+        field_type = field_spec["type"]
+
+        values, is_multi_value = _extract_literal_values(field_type)
+
+        if is_multi_value:  # Multi-value field (literal or free-text)
+            fields[key] = (field_type, Field(default_factory=list))
+        elif values:  # Single-value literal field
+            fields[key] = (field_type, Field(default=values[0]))
+        else:  # Single-value free-text fields: str, int, float, bool
+            type_default_map = {
+                str: "",
+                int: 0,
+                float: 0.0,
+                bool: False,
+            }
+            if field_type in type_default_map:
+                fields[key] = (field_type, Field(default=type_default_map[field_type]))
+            else:
+                msg = f"Unsupported field type for key '{key}': {field_type!r}"
+                raise ValueError(msg)
+
+    model: type[BaseModel] = create_model("DocumentMetadata", __base__=BaseModel, **fields)
+
+    # Base system prompt.
+    system_prompt = (
+        "You are a precise metadata extractor for documents.\n"
+        "Extract the requested metadata from the provided source text.\n"
+        "For constrained fields, only use the provided allowed values.\n"
+        "For free-text fields, provide concise and accurate responses.\n"
```

Comment:
I wonder if we should allow unconstrained extraction for the first version. The reason is that it's not clear to me what we would do with such metadata yet. And from prior experience, we know this is going to yield very noisy metadata.